OpenRA.Mods.Cnc/Traits/World/BuildableAutoTerrainRenderer.cs

```diff
@@ -0,0 +1,206 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2021 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using OpenRA.Graphics;
+using OpenRA.Mods.Common.Terrain;
+using OpenRA.Mods.Common.Traits;
+using OpenRA.Primitives;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Cnc.Traits
+{
+	[TraitLocation(SystemActors.World)]
+	[Desc("Attach this to the world actor. Renders the state of " + nameof(BuildableAutoTerrainLayer))]
+	public class BuildableAutoTerrainRendererInfo : TraitInfo
+	{
+		[Desc("Palette to render the layer sprites in.")]
+		public readonly string Palette = TileSet.TerrainPaletteInternalName;
+
+		public override object Create(ActorInitializer init) { return new BuildableAutoTerrainRenderer(init.Self, this); }
+	}
+
+	public class BuildableAutoTerrainRenderer : IRenderOverlay, IWorldLoaded, ITickRender, IRadarTerrainLayer, INotifyActorDisposing
+	{
+		readonly World world;
+		readonly BuildableAutoTerrainRendererInfo info;
+		readonly BuildableAutoTerrainLayer buildableAutoTerrainLayer;
+		readonly Dictionary<CPos, Sprite> dirty;
+		readonly ITiledTerrainRenderer terrainRenderer;
+		readonly CellLayer<(Color, Color)> radarColor;
+
+		TerrainSpriteLayer terrainSpriteLayer;
+		ITemplatedTerrainInfo terrainInfo;
+		PaletteReference paletteReference;
+
+		public BuildableAutoTerrainRenderer(Actor self, BuildableAutoTerrainRendererInfo info)
+		{
+			this.info = info;
+			world = self.World;
+			dirty = new Dictionary<CPos, Sprite>();
```

Comment:
Why not do this at the declaration?
`readonly Dictionary<CPos, Sprite> dirty = new Dictionary<CPos, Sprite>();`

Reply:
So the lines are evenly long.